File: listme-frontend/src/crdt/ConflictDetector.ts
import { VectorClock } from './VectorClock'
import type { CrdtOperation } from './types'

export interface Conflict {
  a: CrdtOperation
  b: CrdtOperation
}

/**
 * Client-side conflict detector — mirrors backend ConflictDetector logic.
 *
 * Two operations on the same item are concurrent if neither vector clock
 * causally dominates the other.
 */
export function detectConflicts(ops: CrdtOperation[]): Conflict[] {
  const conflicts: Conflict[] = []

  for (let i = 0; i < ops.length; i++) {
    for (let j = i + 1; j < ops.length; j++) {
      const a = ops[i]!
      const b = ops[j]!

      const targetA = a.payload['itemId']
      const targetB = b.payload['itemId']
      if (!targetA || targetA !== targetB) continue

      const clockA = VectorClock.of(a.vectorClock)
      const clockB = VectorClock.of(b.vectorClock)

      if (clockA.compare(clockB) === 'CONCURRENT') {
        conflicts.push({ a, b })
      }
    }
  }

  return conflicts
}

/**
 * LWW (Last-Write-Wins) resolution for two concurrent operations.
 * Falls back to lexicographic deviceId comparison for equal timestamps.
 */
export function resolveLww(a: CrdtOperation, b: CrdtOperation): CrdtOperation {
  const tsA = (a.payload['timestamp'] as number) ?? a.createdAt
  const tsB = (b.payload['timestamp'] as number) ?? b.createdAt

  if (tsA > tsB) return a
  if (tsB > tsA) return b
  // Tiebreak: higher deviceId wins (deterministic across all devices)
  return a.id > b.id ? a : b
}

File: listme-backend/src/main/java/com/oliwier/listmebackend/crdt/ConflictDetector.java
package com.oliwier.listmebackend.crdt;

import com.oliwier.listmebackend.domain.model.CrdtOperation;

import java.util.ArrayList;
import java.util.List;

/**
 * Detects conflicting (concurrent) operations in a list of CRDT operations.
 *
 * Two operations on the same target (same itemId) are concurrent if
 * neither of their vector clocks is causally before the other.
 */
public class ConflictDetector {

    public record Conflict(CrdtOperation a, CrdtOperation b) {}

    /**
     * Returns all pairs of operations that are CONCURRENT on the same item.
     * Caller is responsible for applying merge strategy to resolve them.
     */
    public static List<Conflict> detect(List<CrdtOperation> ops) {
        List<Conflict> conflicts = new ArrayList<>();

        for (int i = 0; i < ops.size(); i++) {
            for (int j = i + 1; j < ops.size(); j++) {
                CrdtOperation a = ops.get(i);
                CrdtOperation b = ops.get(j);

                // Only compare ops that target the same item
                if (!sameTarget(a, b)) continue;

                VectorClock clockA = VectorClock.of(a.getVectorClock());
                VectorClock clockB = VectorClock.of(b.getVectorClock());

                if (clockA.compare(clockB) == VectorClock.Relation.CONCURRENT) {
                    conflicts.add(new Conflict(a, b));
                }
            }
        }
        return conflicts;
    }

    /** Extract itemId from payload for target comparison. */
    private static boolean sameTarget(CrdtOperation a, CrdtOperation b) {
        Object targetA = a.getPayload().get("itemId");
        Object targetB = b.getPayload().get("itemId");
        if (targetA == null || targetB == null) return false;
        return targetA.equals(targetB);
    }
}
